src/OrleansCodeGenerator/RoslynCodeGenerator.cs

```diff
@@ -425,28 +425,44 @@ private static GeneratedSyntax GenerateForAssemblies(List<Assembly> assemblies,
             var module = runtime ? null : type.Module;
             var typeInfo = type.GetTypeInfo();
 
-            // Every type which is encountered must be considered for serialization.
-            if (!typeInfo.IsNested && !typeInfo.IsGenericParameter && typeInfo.IsSerializable)
-            {
-                // If a type was encountered which can be accessed, process it for serialization.
-                if (SerializerGenerationManager.RecordTypeToGenerate(type, module, targetAssembly))
-                {
-                    includedTypes.Add(type);
-                }
-            }
+            // If a type was encountered which can be accessed and is marked as [Serializable], process it for serialization.
+            if (typeInfo.IsSerializable)
+                RecordType(type, module, targetAssembly, includedTypes);
+
+            Type grainStateType;
```

Comment:
I don't think so. The grain class may be marked as `[Serializable]`, yet it may also have a `T` (`grainStateType`) that needs to be recorded. An exotic case but not impossible.